server/WFCORE-4291_restore_legacy_not_graceful_startup_mode.adoc

```diff
@@ -0,0 +1,69 @@
+= WFCORE-4291 - Restore legacy (not "graceful") startup mode
+:author:            Jason Lee
+:email:             jasondlee@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Since Wildfly 11, the server starts up in suspended mode, thus rejecting all incoming requests, until the server has completed booting, at which point the state is set to `RUNNING` and requests are accepted. While this helps ensure that the system is in a stable state before allowing any work to happen, it does have a negative effect on some deployment scenarios reported by Wildfly users. 
+
+One example scenario is this: if Deployment A depends -- at startup -- on external configuration, fetched by an HTTP request to Deployment B, Deployment A will fail to start as the HTTP request for that configuration will be rejected by the starting server.
+
+This RFE proposes a new startup mode, `graceless`, that will allow -- for some subsystems, at least -- requests to be handled much earlier in the application server boot process. In standalone mode, the system administrator can choose to enable this startup mode for that server, and in domain mode, the server group can be configured similarly, applying the configuration choice to all servers in that domain.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-4291[WFCORE-4291]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1320[EAP7-1320]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+* [X] QE
+
+=== Affected Projects or Components
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+For standalone mode, this option will be exposed as a command line option only, with the addition of a `--graceless` option. Since the value is needed before standalone.xml is parsed, it must come from the command line. Given this lack of start mode persistence, system administrators managing standalone instances will need to update their system scripts (`/etc/init.d`, `sysvinit`, `systemd`, etc.)
+
+For domain mode, the configuration will be extended to allow the system administrator to configure how the Host Contoller will start the servers in the group. Every server in the group will be configured to start in the same manner.
+
+=== Nice-to-Have Requirements
+
+=== Non-Requirements
```

Comment:
What about the behavior on mixed domains? We could have server groups with servers running on different versions, in such case, we could have servers supporting the graceless option and servers unsupported them, which could be fine, but maybe it deserves a description about what we are going to do on those cases.

Reply:
+1. This is a standard mixed domain transformation requirement, and should be mentioned.

@yersan Can you suggest wording please? I think that will be helpful to Jason as this is his first experience with this topic.

Reply:
The basic requirement here is that if the graceless attribute is set to true or to an expression (if it supports expressions) then the transformer should reject it. This is the meaning of 'understand and react properly' in this particular case.  A host controller running an earlier version that needs that server group should fail to join the domain in such a config already exists. If such a host controller is already part of the domain and an admin invokes a management op to turn on graceless, that op should fail.

A graceless=false setting should be discarded when sent to the legacy host. It won't understand it, but that's ok because its behavior is the same as a 'false' setting.

^^^ is a description of the general requirement for domain mode when a new config setting is introduced that a host running a prior version cannot understand.

Reply:
Thanks for clearing it up @bstansberry , I was unsure if we could have cases on mixed domains with servers using `--graceless` and servers without using it and I wanted to test it. My missing point was the rejection affects not only when there are slaves registered, but also when a slave is being registered the first time in the domain.

Reply:
> The basic requirement here is that if the graceless attribute is set to true or to an expression (if it supports expressions) then the transformer should reject it. This is the meaning of 'understand and react properly' in this particular case. A host controller running an earlier version that needs that server group should fail to join the domain 

Is there additional work to be done on the transformer, then? I don't recall doing anything like that...

Reply:
@jasondlee Yes, I think so, as a guide you can take a look at how `socket-binding-default-interface` is registered in [ServerGroupTransformers](https://github.com/wildfly/wildfly-core/blob/385084ff0a68558513502c7c55f578cc592c1b5c/host-controller/src/main/java/org/jboss/as/domain/controller/transformers/ServerGroupTransformers.java#L63-L67)

Reply:
Transformers updated. Tests passing (thanks @bstansberry)

Reply:
Thanks @jasondlee. Can you please describe the behavior @yersan mentions in the first comment in this analysis doc?